pkg/codegen/pcl/program.go

```diff
@@ -169,6 +185,35 @@ func (p *Program) Source() map[string]string {
 	return source
 }
 
+// SourceFiles returns an in-memory representation of all files used to construct a Program,
+// including source files of used components
+func (p *Program) SourceFiles(directory string) *ProgramDirectory {
+	entries := make([]*ProgramFileOrDirectory, 0)
+	for _, file := range p.files {
+		entries = append(entries, &ProgramFileOrDirectory{
+			File: &ProgramFile{
+				FileName: file.Name,
+				Content:  file.Bytes,
+			},
+		})
+	}
+
+	for _, node := range p.Nodes {
+		switch node := node.(type) {
+		case *Component:
+			componentDirectory := filepath.Join(directory, node.source)
```

Comment:
We might see the same `source` multiple times.
Also `source` is a path, this should reconstruct the ProgramDirectory with each named part, not just the overall path.